# Basic :

`Not normal basics but basics covered by actual datasets not basic numopy arrays. It will be quite difficult to understand at first. But Trust me THis will prove to be very beneficial for you in future. `


## We will start this notebook with tensorOPS with irirs datset:


scalars (single numbers), vectors (1D arrays), and matrices (2D arrays). These are the fundamental building blocks of all deep learning computations.

In PyTorch, the fundamental data structure that encompasses all of these is the `torch.Tensor`.

* A scalar is a 0-dimensional tensor.
* A vector is a 1-dimensional tensor.
* A matrix is a 2-dimensional tensor.
* A stack of matrices (like an RGB image) is a 3-dimensional tensor.

A batch of images is a 4-dimensional tensor.
Think of a torch.Tensor as a super-powered NumPy array. It can live on a GPU for massive acceleration and seamlessly works with PyTorch's automatic differentiation engine, which we'll cover soon.

Today, we'll convert the Iris dataset, which is stored in NumPy arrays, into these powerful PyTorch Tensors.

In [5]:
import torch
import sklearn
from sklearn.datasets import load_iris
import numpy as np


print(f"Pytorch version : { torch.__version__ }")
print(f"sklearn - version : {sklearn.__version__}")



################################################
# Load the dataset

iris = load_iris()
X = iris.data
y = iris.target

print(f"X.datatype : {X.dtype}, y.datatype : {y.dtype}")


############################################
# Convertour dataset to pytorch tensors




# X is our feature matrix
# Neural networks typically expect input data to be float

X_tensor = torch.tensor(X, dtype=torch.float32)

# y is our label vector
# Labels for classification are often long integers
# Hence, the dtype 'long'
y_tensor = torch.tensor(y, dtype=torch.long)


# Let's see what we've created.
print("--- Tensor Properties ---")
print(f"Shape of features tensor: {X_tensor.shape}")
print(f"Shape of labels tensor: {y_tensor.shape}\n")

print(f"Data type of features tensor: {X_tensor.dtype}")
print(f"Data type of labels tensor: {y_tensor.dtype}\n")

# .device tells us if the tensor is on the CPU or a GPU. For now, it's CPU.
print(f"Device tensor is stored on: {X_tensor.device}\n")





# --- 4. See the Data ---
# Let's look at the first 5 rows of our data, now as a PyTorch Tensor.
print("--- First 5 Samples ---")
print("Features (X):")
print(X_tensor[:5])
print("\nLabels (y):")
print(y_tensor[:5])
print(f"Which correspond to flower types: {[iris.target_names[i] for i in y_tensor[:5]]}")


Pytorch version : 2.5.1
sklearn - version : 1.6.1
X.datatype : float64, y.datatype : int32
--- Tensor Properties ---
Shape of features tensor: torch.Size([150, 4])
Shape of labels tensor: torch.Size([150])

Data type of features tensor: torch.float32
Data type of labels tensor: torch.int64

Device tensor is stored on: cpu

--- First 5 Samples ---
Features (X):
tensor([[5.1000, 3.5000, 1.4000, 0.2000],
        [4.9000, 3.0000, 1.4000, 0.2000],
        [4.7000, 3.2000, 1.3000, 0.2000],
        [4.6000, 3.1000, 1.5000, 0.2000],
        [5.0000, 3.6000, 1.4000, 0.2000]])

Labels (y):
tensor([0, 0, 0, 0, 0])
Which correspond to flower types: ['setosa', 'setosa', 'setosa', 'setosa', 'setosa']


# Explorations for you to do !!!!!

`Make sure you do these.`



**Beginner**: In the code, change dtype=torch.float32 to dtype=torch.float64 for X_tensor. Rerun the script. What changes in the output? (This explores data precision).

**Intermediate**: The current y_tensor has labels like 0, 1, 2. For some loss functions, you'll need "one-hot encoding" (e.g., 2 becomes [0, 0, 1]). Can you find a PyTorch function to convert y_tensor into a one-hot encoded tensor of shape [150, 3]? (Hint: look into torch.nn.functional.one_hot).

**Advanced**: PyTorch tensors and NumPy arrays can share the same underlying memory location to save memory and avoid copying. Use torch.from_numpy() instead of torch.tensor() to create X_tensor. Now, modify an element in the original X NumPy array (e.g., X[0, 0] = 99.9). Does the X_tensor also change? Why is this a powerful feature, and when could it be dangerous?

#### 5. Dataset-Specific Exercises
Now it's your turn to write the code.

* Replication: Reproduce the exact example above, but use the sklearn.datasets.load_wine dataset. Print the shape and data 
types of the resulting tensors.

* Modification: Use the sklearn.datasets.load_digits dataset. This dataset contains small 8x8 images of handwritten digits. What is the shape of its feature tensor? How does this shape relate to the 8x8 images?

* Creation: Imagine you have a new iris flower with the following measurements: sepal length=5.1, sepal width=3.5, petal length=1.4, petal width=0.2. Create a new 1D tensor from scratch that represents these features using torch.tensor(). What is its shape?

#### 6. Professional Best Practices Introduced
* Data Types Matter: We explicitly set torch.float32 for features and torch.long for labels. This is standard practice and prevents common errors later on. float32 is the default for model weights and is a good balance of precision and memory usage.

* Know Your Data's Home: The .device attribute is critical. All computations between tensors must happen on the same device (e.g., all on the CPU or all on a specific GPU). We'll master this later.

* Reproducibility (A Teaser): For truly reproducible results, you should set a random seed at the beginning of your script. Add torch.manual_seed(42) to the top of the script. We'll see why this is crucial when we start training models.


# ______________________________________________________________